src/Interactive/EditorFeatures/Core/Completion/CompletionProviders/ReplCommandCompletionProvider.cs

```diff
@@ -0,0 +1,63 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Editor.Implementation.Interactive;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.InteractiveWindow;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Utilities;
+using Microsoft.VisualStudio.InteractiveWindow.Commands;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.Completion.CompletionProviders
+{
+    [ExportCompletionProvider("ReplCommandCompletionProvider", LanguageNames.CSharp)]
+    [TextViewRole(PredefinedInteractiveTextViewRoles.InteractiveTextViewRole)]
+    [Order(Before = PredefinedCompletionProviderNames.Keyword)]
+    internal abstract class ReplCompletionProvider : CompletionListProvider
+    {
+        protected abstract Task<TextSpan> GetTextChangeSpanAsync(Document document, int position, CancellationToken cancellationToken);
+        protected abstract bool ShouldDisplayCommandCompletions(SyntaxTree tree, int position, CancellationToken cancellationToken);
+
+        public override async Task ProduceCompletionListAsync(CompletionListContext context)
+        {
+            var document = context.Document;
+            var position = context.Position;
+            var cancellationToken = context.CancellationToken;
+
+            // the provider might be invoked in non-interactive context:
+            Workspace ws;
+            if (Workspace.TryGetWorkspace(document.GetTextAsync(cancellationToken).WaitAndGetResult(cancellationToken).Container, out ws))
```

Comment:
no. This is an async method, we should await and add a ConfigureAwait(false) at the end).